src/main/java/se/liu/ida/hefquin/engine/data/impl/VocabularyMappingImpl.java

```diff
@@ -40,23 +54,293 @@ public VocabularyMappingImpl(final Set<Triple> triples) {
+				boolean allSubPatternsAreTriplePatterns = true; // assume yes
+
+				for( final TriplePattern m : ((BGP) j).getTriplePatterns() ) {
+					final SPARQLGraphPattern p = translatePredicate(m);
+					allSubPatterns.add(p);
+
+					if ( allSubPatternsAreTriplePatterns && p instanceof TriplePattern ) {
+						tpSubPatterns.add( (TriplePattern) p );
+					}
+					else {
+						allSubPatternsAreTriplePatterns = false;
+					}
+				}
+
+				if ( allSubPatternsAreTriplePatterns ) {
+					predicateTranslation.add( new BGPImpl(tpSubPatterns) );
+				}
+				else {
+					predicateTranslation.add( new SPARQLGroupPatternImpl(allSubPatterns) );
+				}
+
+			} else if (j instanceof TriplePattern) {
+				predicateTranslation.add(translatePredicate((TriplePattern) j));
+				
+			} else {
+				throw new IllegalArgumentException(j.getClass().getName());
+			}
+		}
+		
+		if(predicateTranslation.size() > 1) {
+			final SPARQLUnionPattern translation = new SPARQLUnionPatternImpl(predicateTranslation);
+			return translation;
+		} else {
+			return predicateTranslation.get(0);
+		}
 	}
 	
 	protected Set<TriplePattern> translateSubject(final TriplePattern tp){
-		return null;
+		Set<TriplePattern> results = new HashSet<TriplePattern>();
+		org.apache.jena.graph.Triple t = tp.asJenaTriple();
+		if (t.getSubject().isVariable()) {
+			results.add(tp);
+			return results;
+		}
+		final Node p = NodeFactory.createVariable("p");
+		final Node o = NodeFactory.createVariable("o");
+		final TriplePattern tpQuery = new TriplePatternImpl(t.getSubject(), p, o);
+		for (final Triple m : getMappings(tpQuery)) {
+			if (m.asJenaTriple().getPredicate().getURI().equals(OWL.sameAs.getURI())) {
+				final TriplePattern translation = new TriplePatternImpl(m.asJenaTriple().getObject(), t.getPredicate(), t.getObject());
+				results.add(translation);
+			} else {
+				throw new IllegalArgumentException(m.asJenaTriple().getPredicate().getURI());
+			}
+		}
+		return results;
 	}
 	
 	protected SPARQLGraphPattern translateObject(final TriplePattern tp){
-		return null;
+		org.apache.jena.graph.Triple t = tp.asJenaTriple();
+		if(!t.getPredicate().isURI()) {
+			//System.out.print("Not URI : " + t.getPredicate().toString());
+			return tp;
+		} else {
+			if(!t.getPredicate().getURI().equals(RDF.type.getURI())) {
+				//System.out.print("Not RDF type : " + t.getPredicate().getURI());
+				return tp;
+			}
+		}
+		final Node p = NodeFactory.createVariable("p");
+		final Node o = NodeFactory.createVariable("o");
+		final TriplePattern tpQuery = new TriplePatternImpl(t.getObject(), p, o);		
+		List<SPARQLGraphPattern> resultsList = new ArrayList<SPARQLGraphPattern>();
+		for (final Triple m : getMappings(tpQuery)) {
+			String predicate = m.asJenaTriple().getPredicate().getURI();
+			
+			if(predicate.equals(OWL.sameAs.getURI())) {
+				final TriplePattern translation = new TriplePatternImpl(t.getSubject(), t.getPredicate(), m.asJenaTriple().getObject());
+				resultsList.add(translation);
```

Comment:
Related to my comment above, it seems that you are indeed aiming to using owl:sameAs mappings for the objects.